test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs

```diff
@@ -584,6 +523,360 @@ public void GetApiDescription_IncludesResponseFormats_FilteredByType()
+            Assert.Same(ApiParameterSource.Query, quantity.Source);
+            Assert.Equal(typeof(int), quantity.Type);
+
+            var product = Assert.Single(description.ParameterDescriptions, p => p.Name == "Product");
+            Assert.Same(ApiParameterSource.Query, product.Source);
+            Assert.Equal(typeof(OrderProductDTO), product.Type);
+        }
+
+        [Fact]
+        public void GetApiDescription_ParameterDescription_BreaksCycles()
+        {
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsCycle));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(1, description.ParameterDescriptions.Count);
+
+            var c = Assert.Single(description.ParameterDescriptions);
+            Assert.Same(ApiParameterSource.Query, c.Source);
+            Assert.Equal("C.C", c.Name);
+            Assert.Equal(typeof(Cycle1), c.Type);
+        }
+
+        [Fact]
+        public void GetApiDescription_ParameterDescription_DTOWithCollection()
+        {
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsHasCollection));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(1, description.ParameterDescriptions.Count);
+
+            var products = Assert.Single(description.ParameterDescriptions);
+            Assert.Same(ApiParameterSource.Query, products.Source);
+            Assert.Equal("Products", products.Name);
+            Assert.Equal(typeof(Product[]), products.Type);
+        }
+
+        // If a property/parameter is a collection, we automatically treat it as a leaf-node.
+        [Fact]
+        public void GetApiDescription_ParameterDescription_DTOWithCollection_ElementsWithBinderMetadataIgnored()
+        {
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsHasCollection_Complex));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(1, description.ParameterDescriptions.Count);
+
+            var c = Assert.Single(description.ParameterDescriptions);
+            Assert.Same(ApiParameterSource.ModelBinding, c.Source);
+            Assert.Equal("c", c.Name);
+            Assert.Equal(typeof(HasCollection_Complex), c.Type);
+        }
+
+        [Fact]
+        public void GetApiDescription_ParameterDescription_RedundentMetadataMergedWithParent()
```

Comment:
Two parameters. We never partially describe a model. Adding a test for this case.